llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -1543,6 +1543,29 @@ static void computeKnownBitsFromOperator(const Operator *I,
         }
         break;
       }
+
+      case Instruction::UDiv: {
+        // For UDiv, the result can never exceed either the numerator, or the
+        // start value, whichever is greater. The case where the PHI is not
+        // the numerator of the UDiv is already handled by other code.
+        if (BO->getOperand(0) != P)
+          break;
+        [[fallthrough]];
+      }
+
+      case Instruction::URem: {
+        // For URem, the result can never exceed the start value.
+        SimplifyQuery RecQ = Q.getWithoutCondContext();
+
+        unsigned OpNum = P->getOperand(0) == R ? 0 : 1;
+        Instruction *RInst = P->getIncomingBlock(OpNum)->getTerminator();
+
+        RecQ.CxtI = RInst;
+        computeKnownBits(R, DemandedElts, Known2, Depth + 1, RecQ);
+        Known.Zero.setHighBits(Known2.countMinLeadingZeros());
```

Comment:
is `urem` where `phi` is on lhs also handled already?

Reply:
Is there a reason you don't also try to get information from `lhs`. For `udiv` and `urem` can't exceed `lhs` either.

Reply:
Isn't the BTC of the loop unknown (since we can't use SCEV), and hence is possibly-zero? In this case, will the PHI not take the initial value?

Reply:
I don't follow, if its zero and we get wrong analysis is fine to handle the UB however we like. If its non-zero the analysis holds no?

Reply:
Consider the following loop:

```llvm
loop:
  %iv = phi i64 [9, %entry], [%iv.next, %loop]
  %iv.next = urem i64 3, %iv
  %exitcond = icmp eq %iv.next, 3
  br i1 %exitcond, label %exit, label %loop
```

It has trip-count = 1, backedge-taken-count = 0. It is not UB. Do you agree that the maximum value of %iv is 9, and not 3?

Reply:
How does that differ from RHS vs LHS?

Reply:
I don't check RHS or LHS explicitly at all:

```cpp
        RecQ.CxtI = RInst;
        computeKnownBits(R, DemandedElts, Known2, Depth + 1, RecQ);
        Known.Zero.setHighBits(Known2.countMinLeadingZeros());
```

Here, `RInst` and `R` correspond to the PHI's start value.

Reply:
Oh I see, I got myself a bit confused. This makes sense. Sorry for the confusion.